### PR TITLE
Major change to fix issues with React Hooks

### PR DESCRIPTION
**Pull request by justin808**
See details in https://github.com/shakacode/react_on_rails/issues/1198

# BREAKING CHANGE

I’m updating React on Rails. The new API for **ReactOnRails.register** will take an object which registers either

1. React functional or class components
2. render-functions, as defined below.

The update concerns render-functions (formerly called "generator functions"). Previously, generator functions would return JSX and there were sort of like a React function component. But not quite, as you could not use React Hooks within them.

With this update, render-functions need to return either a

1. React Function Component
2. React Class Component
3. An object representing server-side rendering output

***Note, you can no longer return JSX!***

If you use React component or create a function that returns a React component, then the React on Rails library will call React.createElement and then call ReactDOM.render (or hydrate).

This ensures that React Hooks can always be used with both the React functional or class components and render-functions.

In the case of the render-function, if you could just return the React element (or JSX) and you used the hook API, then you’d get an obscure “hooks don’t work” error.

Use regular React functional or class component or a render-function for your client-side bundle:

```javascript
// React functional component
export default (props) => (
  <App {...props} />
);
```

Or a render-function. Note you can't return just the JSX (React element), but you need to return either a React functional or class component.
```javascript
// React functional component
export default (props, railsContext) => (
  () => <App {{railsContext, ...props}} />
);
```

Note, this doesn't work, because this function just returns a React Element
```javascript
// React functional component
export default (props, railsContext) => (
  <App {{railsContext, ...props}} />
);
```

While this does result in changing all the cases that looked like a React functional component, but had 2 params, I do believe this is worth it.

#1268  Performance? 

In my update to React on Rails, I’m changing some APIs. I’m wondering if calling a function to create a function that has a closure with say very a large props JS object, and then returning that function and then
the function will again be called by React.createElement(MyComponent, props)…So the props passed to the React function are not really necessary b/c they can be in a closure.

I’m concerned that having the props passed to the real React component will be bad for performance if they already should be used from the closure. However, **JavaScript passes objects by reference, so I don't see any issue.**

This could be optimized by not passing any props (which would go into unused param _props below). However, that’s going to be confusing in some cases. I can see people defining the inner component using the same name props and that would shadow the outer one.

```js
import React from 'react';
const MyAppComponent = (props, railsContext) => (
  // NOTE: need to wrap in a function so this is proper React function component that can use
  // hooks
  // the props get passed again, but we ignore since we use a closure
  // or should we
  (_props) =>
      <div>
        <p>props are: {JSON.stringify(props)}</p>
        <p>railsContext is: {JSON.stringify(railsContext)}
        </p>
      </div>
);
export default MyAppComponent;
```

---
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1268)

### CODE REVIEW

File: spec/dummy/client/app/startup/DeferredRenderAppServer.jsx [hunk 1]
```diff
@@ -34,7 +34,7 @@ const DeferredRenderAppServer = (_props, railsContext) => {
     return { error, redirectLocation };
   }
 
-  return <RouterContext {...routerProps} />;
+  return () => <RouterContext {...routerProps} />;
```
Comment (justin808) [inline spec/dummy/client/app/startup/DeferredRenderAppServer.jsx:37]: This makes sense, @Judahmeek. It needs a React function component, not JSX.

File: spec/dummy/spec/system/integration_spec.rb [hunk 2]
```diff
@@ -66,10 +66,10 @@ def finished_all_ajax_requests?
 
     context "React Hooks" do
       context "Simple stateless component" do
-        include_examples "React Component", "div#HelloWorldApp-react-component-6"
+        include_examples "React Component", "div#HelloWorld-react-component-6"
       end
       context "render function that takes props" do
-        include_examples "React Component", "div#HelloWorldApp-react-component-7"
+        include_examples "React Component", "div#HelloWorld-react-component-7"
```
Comment (justin808) [inline spec/dummy/spec/system/integration_spec.rb:72]: An idea on why this changed?

**Comment by justin808** [comment]
Only pending thing for merge is to double check the CHANGELOG, README, and upgrade docs.

File: docs/basics/upgrading-react-on-rails.md [hunk 3]
```diff
@@ -5,6 +5,54 @@ If you would like help in migrating between React on Rails versions or help with
 
 We specialize in helping companies to quickly and efficiently move from versions before 9 to current. The older versions use the Rails asset pipeline to package client assets. The current and recommended way is to use Webpack 4 for asset preparation. You may also need help migrating from the `rails/webpacker`'s Webpack configuration to a better setup ready for Server Side Rendering.
 
+## Upgrading to v13
+* Make sure that are are on a relatively more recent version of rails and webpacker.
```
Comment (tahsin352) [inline docs/basics/upgrading-react-on-rails.md]: There is extra `are`. Please remove.

File: docs/contributor-info/errors-with-hooks.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+npm ERR! invalid: react-on-rails@12.0.0 /Users/justin/shakacode/react-on-rails/react_on_rails/spec/dummy/node_modules/react-on-rails
+npm ERR! extraneous: react@16.13.1 /Users/justin/shakacode/react-on-rails/react_on_rails/spec/dummy/node_modules/react-on-rails/node_modules/react
+```
+
+Make sure there is only one version of React installed!
+
+If you used yarn link, then you'll have two versions of React installed.
+
+Instead use Yalc.
```
Comment (tahsin352) [inline docs/contributor-info/errors-with-hooks.md]: cool.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -97,16 +97,26 @@ In addition to testing the Ruby parts out, you can also test the node package pa
 ```sh
 cd react_on_rails/
 yarn
-yarn run build
-yarn install-react-on-rails
+yarn run build:watch
+yalc pubish react-on-rails
```
Comment (tahsin352) [inline CONTRIBUTING.md]: `publish`

File: spec/dummy/spec/system/integration_spec.rb (hunk 2 above)
Comment (Judahmeek) [reply spec/dummy/spec/system/integration_spec.rb:72]: Those are the Ids that you used in your initial commit: https://github.com/shakacode/react_on_rails/pull/1268/commits/be37b3410de1974f16066951bc41fa73fa6f3c61?file-filters%5B%5D=.erb#diff-13675373e6eaa2b6e91995c3ec8974f7

**Comment by justin808** [comment]
Thanks for the help @Judahmeek!

**Comment by jacksonrayhamilton** [comment]
Glad to see this merged!

Should we expect React on Rails v12 to be released soon?

**Comment by simalexan** [comment]
@justin808 this is great, appreciate your work! 
Just wondering as @jacksonrayhamilton when should we see this published?

**Comment by justin808** [comment]
OMG! Your thank you messages really fire me up! I'll try to get this one by this weekend!

BTW, if any of you can consider even the tiniest sponsorship, that would be so appreciated!

https://github.com/sponsors/shakacode

File: CONTRIBUTING.md (hunk 5 above)
Comment (justin808) [reply CONTRIBUTING.md]: Thank you @tahsin352 for your review!!!!

@Judahmeek did we get all of these?

**Pull request merged**